doc/developer/design/20240205_cluster_specific_optimization.md

```diff
@@ -0,0 +1,283 @@
+    optimizer feature flags using the `EXPLAIN WITH(...)` syntax
+    (MaterializeInc/materialize#22744). There are some known issues with the
+    current state of the code, but we have fixes for those that should be merged
+    soon (pending review).
+2.  **Assessing the quality of the running dataflows.** This is needed because
+    sometimes the impact of an optimizer change is not directly evident in the
+    optimized plan. Instead, we need to deploy and run the modified dataflow and
+    compare its performance against its current version.
+
+The current document proposes tooling and workflows in aid of (2).
+
+## Success Criteria
+
+<!--
+What does a solution to this problem need to accomplish in order to
+be successful?
+
+The criteria should help us verify that a proposed solution would solve
+our problem without naming a specific solution. Instead, focus on the
+outcomes we hope result from this work. Feel free to list both qualitative
+and quantitative measurements.
+-->
+
+-   Members of the compute team have agreed upon the developed the minimal amount of
+    tooling and code infrastructure required to enable the evaluation of proposed
+    optimizer changes in production.
+
+-   The new feature has been used to asses the potential impact of the following
+    optimizer changes:
+    -   [optimizer: Flag for controlling join
+        planning (#23318)](https://github.com/MaterializeInc/materialize/pull/23318)
+    -   [DNM: Draft for variadic outer join
+        lowering (#24345)](https://github.com/MaterializeInc/materialize/pull/24345)
+
+## Out of Scope
+
+<!--
+What does a solution to this problem not need to address in order to be
+successful?
+
+It's important to be clear about what parts of a problem we won't be solving
+and why. This leads to crisper designs, and it aids in focusing the reviewer.
+-->
+
+Things that will be nice to have, but are intentionally left out of scope from
+the current proposal are listed below with a brief motivation for that decision.
+
+-   **Not requiring `mz_system` access in order to do a runtime validation
+    experiment.** This doesn't seem urgent because we will always use
+    `mzexplore` to first identify a small set of candidate dataflows. For those,
+    we will in turn run a dedicated experiment together with a designated
+    engineer that can temporarily request `mz_system` privileges with Teleport
+    (similar to the `lgalloc` experiments).
+-   **Provide tooling to automatically duplicate dataflow-backed catalog items
+    from a reference cluster to a temporary cluster used to run an experiment.**
+    While this might be very helpful in general, we believe that most of our
+    customers deploy their DDL statements with a script. For the first iteration
+    of this feature it should be therefore sufficient to ask them to run a
+    modified version of this script (with changed catalog item names) against
+    the temporary cluster while running our experiments.
```

Comment:
I worry a bit that this will be exhausting for y'all. But, I think it properly puts the "M" in "MVP", and any tooling to easily "clone" an existing cluster will be easy enough to build on top of this proposal.

Reply:
I'd also like to point out that this will make it possible for unresponsive customers to block our testing, which will be frustrating. This, together with the danger of customer accidentally breaking their env by running the wrong script, makes me think we should maybe try to get something that works without customer interaction for the MVP. I think it can be as simple as a runbook that describes how to use `mzexplore` to create the appropriate SQL.

Reply:
I'll start prototyping the necessary changes in Python when I run the first batch of experiments. Hopefully when I'm done with them I'll have a good idea what references where need to be updated.